utils/generate_protocol_types.py

```diff
@@ -0,0 +1,450 @@
+                return d[key]
+
+        raise KeyError(f'{", ".join([str(s) for s in k])} all not found in {d}')
+
+    def convert_js_to_py_type(self, item_info: Union[Dict[str, Any], str], domain_name) -> str:
+        """
+        Generates a valid python type from the JS type. In the case of type_info being a str, we simply return the
+        matching python type from self.js_to_py_types. Otherwise, in the case of type_info being a Dict, we know that
+        it will contain vital information about the type we are trying to convert.
+
+        The domain_name is used to qualify relative forward reference in type_info. For example, if
+        type_info['$ref'] == 'foo', domain_name would be used produce an absolute forward reference, ie domain_name.foo
+
+        :param item_info: Dict or str containing type_info
+        :param domain_name: path to resolve relative forward references in type_info against
+        :return: valid python type, either in the form of an absolute forward reference (eg Protocol.bar.foo) or
+            primitive type (eg int, float, str, etc)
+        """
+        if isinstance(item_info, str):
+            _type = self.js_to_py_types[item_info]
+        elif 'items' in item_info:
+            assert item_info['type'] == 'array'
+            if '$ref' in item_info['items']:
+                ref = item_info["items"]["$ref"]
+                _type = f'List[{self.get_forward_ref(ref, domain_name)}]'
+            else:
+                _type = f'List[{self.convert_js_to_py_type(item_info["items"]["type"], domain_name)}]'
+        else:
+            if '$ref' in item_info:
+                _type = self.get_forward_ref(item_info["$ref"], domain_name)
+            else:
+                if 'enum' in item_info:
+                    _enum_vals = ", ".join([f'\'{x}\'' for x in item_info["enum"]])
+                    _type = f'Literal[{_enum_vals}]'
+                else:
+                    _type = self.js_to_py_types[item_info['type']]
+
+        return _type
+
+    @staticmethod
+    def get_forward_ref(relative_ref: str, potential_domain_context: str):
+        """
+        Generates a forward absolute forward reference to Protocol class attr. If the reference is relative
+        to a nested class, the full path is resolved against potential_domain_context. In the case of
+        the reference being relative to the Protocol class, the path is simple resolved against the Protocol class
+
+        :param relative_ref: reference to another class, in the form of foo or foo.bar
+        :param potential_domain_context: context to resolve class against if relative_ref is relative to it
+        :return: absolute forward reference to nested class attr
+        """
+        if len(relative_ref.split('.')) == 2:
+            non_fw_ref = f'Protocol.{relative_ref}'
+        else:
+            non_fw_ref = f'Protocol.{potential_domain_context}.{relative_ref}'
+        return f'\'{non_fw_ref}\''
+
+
+class TypingCodeGenerator:
```

Comment:
Nit Pick: Some of these brain classes could use some doctrings since they do quite a bit!

Reply:
I will try to add some